### PR TITLE
exp326: curated cCRE/enhancer datasets + training script (de-contaminate the v4 enhancer arm)

### PR DESCRIPTION
**Pull request by gonzalobenegas**
🤖 Implements the dataset + training scaffolding for the **#326** curated-enhancer experiment — two de-contaminated derivations of the v4 `ccre_non_promoter` arm to test whether removing exon-overlap (#283 **H1**) and non-enhancer cCRE heterogeneity (#283 **H2**) unsticks distal (enhancer) VEP. No experiment is run here; this is the code to build the datasets + train the arms.

## What this adds

Two new `intervals_versions`, both sub-filters of `v4_ccre_non_promoter` (NOT part of the six-way v4 partition):

| dataset | filter | tests |
|---|---|---|
| `v4_ccre_noexon` | ccre windows with **zero** overlap of any other functional element (CDS / 3′UTR / ncRNA exon / TSS+5′UTR), via the disjoint v4 `*_frac` columns | H1 contamination |
| `v4_ccre_noexon_enhancer` | the above **+** enhancer-dominant (dELS+pELS bp ≥ other non-PLS cCRE bp), population-matched to `val_enhancer` | H1 + H2 |

Falsifiable predictions, read against exp232's existing heatmap (the arms are held byte-identical to exp232 except `TRAIN_DATASETS`): **(P1)** the curated arm's off-diagonal splicing/synonymous AUPRC collapses toward the 0.10 floor (the coding leak was CDS contamination); **(P2)** distal AUPRC lifts off ~0.127 and the `val_enhancer` LL-gap↔AUPRC correlation flips from −0.10 to positive.

## Changes

**Library** (testable home) — `src/marin_dna/pipelines/zoonomia_projection_dataset/region_labels.py`
- [`select_ccre_noexon`](https://github.com/Open-Athena/marin-dna/blob/73cd04a0d42ce419fc9fef3d01f809ddef0959ea/src/marin_dna/pipelines/zoonomia_projection_dataset/region_labels.py#L516) — arm A filter.
- [`select_ccre_noexon_enhancer`](https://github.com/Open-Athena/marin-dna/blob/73cd04a0d42ce419fc9fef3d01f809ddef0959ea/src/marin_dna/pipelines/zoonomia_projection_dataset/region_labels.py#L554) — arm B filter (joins the cCRE parquet for per-class membership; assumes `ccre_flank=0`, the v4 setting).
- [`write_curated_ccre_hf_readme`](https://github.com/Open-Athena/marin-dna/blob/73cd04a0d42ce419fc9fef3d01f809ddef0959ea/src/marin_dna/pipelines/zoonomia_projection_dataset/region_labels.py#L1176) — minimal provenance card.
- 6 new tests; the full `test_region_labels.py` suite is **40/40 green**.

**Pipeline** — `snakemake/zoonomia_projection_dataset`
- [`derive_subset_v4_ccre_noexon`](https://github.com/Open-Athena/marin-dna/blob/73cd04a0d42ce419fc9fef3d01f809ddef0959ea/snakemake/zoonomia_projection_dataset/workflow/rules/regions.smk#L348) / [`_enhancer`](https://github.com/Open-Athena/marin-dna/blob/73cd04a0d42ce419fc9fef3d01f809ddef0959ea/snakemake/zoonomia_projection_dataset/workflow/rules/regions.smk#L366) emit the `query_names` lists from the v4 region-labels parquet (+ cCRE / defined for B).
- [`dataset_hf_readme_curated_ccre`](https://github.com/Open-Athena/marin-dna/blob/73cd04a0d42ce419fc9fef3d01f809ddef0959ea/snakemake/zoonomia_projection_dataset/workflow/rules/dataset.smk#L269) + a `ruleorder` so the minimal card wins over the v4 partition card (which would fail looking these up in the six-label composition TSV).
- `config.yaml`: both registered in `intervals_versions`.
- [`sky/upload.yaml`](https://github.com/Open-Athena/marin-dna/blob/73cd04a0d42ce419fc9fef3d01f809ddef0959ea/snakemake/zoonomia_projection_dataset/sky/upload.yaml#L38): new `UPLOAD_MODE=curated_ccre` (targets only the two) + the 3 rules whitelisted in `ALLOWED_RULES`.

**Training** — [`experiments/exp326_curated_enhancers.py`](https://github.com/Open-Athena/marin-dna/blob/73cd04a0d42ce419fc9fef3d01f809ddef0959ea/experiments/exp326_curated_enhancers.py): clone of `exp232_per_region.py` with only `TRAIN_DATASETS` swapped (geometry, optimizer, 5K×8192, val recipes incl. `val_enhancer`, mendelian eval, `v5p-8` all verbatim). `SWEEP_DATASETS` selects the arm.

**Docs** — pipeline README documents both datasets + the curation filters + the `curated_ccre` upload mode.

## Validation

<details><summary>query_names dry-run plans exactly the two new rules (no projection cascade); upload chain wires both datasets through the curated card (not the v4 card)</summary>

```
$ snakemake -n -- results/projection/min0.20/subsets_def/v4_ccre_noexon{,_enhancer}.query_names.txt
job                                      count
derive_subset_v4_ccre_noexon                 1
derive_subset_v4_ccre_noexon_enhancer        1
total                                        2

# full upload chain (…/upload.done/zoonomia-v1-v4_ccre_noexon{,_enhancer}):
subset_dataset_derived 2 · dataset_hf_readme_curated_ccre 2 · prepare_training_shards 2 ·
compress_shard 128 · hf_upload_dataset 2   (+ the known local()-FASTA projection
cascade, collapsed at build time by sky/upload.yaml's --allowed-rules)
```
</details>

- ruff + ruff-format clean on all changed files. (pre-commit `mypy` reports 18 **pre-existing** errors in `evals/leaderboard.py` + `evals/sge.py` — files untouched here; CI skips mypy.)

## Out of scope / next steps (run separately)

- **Build + upload** the two HF datasets: `sky launch … sky/upload.yaml --env UPLOAD_MODE=curated_ccre --env COMMIT_SHA=$(git rev-parse HEAD)`.
- **Train** the two arms via iris (one `SWEEP_DATASETS` per job), then read the distal column + P1/P2 vs exp232.

#326 stays open after this merges — it tracks the experiment's *result*, which lands once the arms run. Refs #283 (the H1/H2 diagnosis), #232 (the per-region run this is held identical to).

### CODE REVIEW

File: experiments/exp326_curated_enhancers.py [hunk 1]
```diff
@@ -0,0 +1,552 @@
...
+
+# Five region-specific validation recipes from PR #171, tokenized functional +
+# nonfunctional for the LL-gap signal. Drops val_utr5 / val_promoter (both
+# subsumed by the gene-centric ±255 bp val_tss_pc); keeps val_enhancer
+# (chromatin-side, not subsumed).
+VAL_DATASETS: tuple[tuple[str, str], ...] = (
+    ("val_cds", "bolinas-dna/zoonomia-v1-val_cds"),
+    ("val_utr3", "bolinas-dna/zoonomia-v1-val_utr3"),
+    ("val_ncrna", "bolinas-dna/zoonomia-v1-val_ncrna"),
+    ("val_enhancer", "bolinas-dna/zoonomia-v1-val_enhancer"),
```
Comment (gonzalobenegas) [inline experiments/exp326_curated_enhancers.py:139]: 🤖 ~~Review (design note)~~ **Withdrawn — no leakage** (clarified by @gonzalobenegas). The `val_enhancer` LL gap is self-supervised conservation (`LL(conserved) − LL(non-conserved)` from phyloP case-encoding) — it uses no variant/pathogenicity/allele-frequency info, so nothing leaks into the held-out distal AUPRC (different label space; the model having seen the enhancer sequence can't leak a variant label it never saw). The train-region↔val-region pairing is the intended matched-pair design for **every** arm (cds↔val_cds, ncrna↔val_ncrna, …) — the other 7 subsets show 0.78–0.98 LL↔AUPRC under the same pairing — so population-matching this arm to `val_enhancer` just sharpens that pair, it doesn't add a confound. Original note below for the record.

---

_Original:_ the enhancer arm is population-matched to `val_enhancer`; consider whether the probe anchors overlap the training windows. (Superseded by the above.)

File: snakemake/zoonomia_projection_dataset/workflow/rules/dataset.smk [hunk 2]
```diff
@@ -70,7 +79,9 @@ for _d in SPECIES_SUBSET_DATASETS:
     SPECIES_SUBSET_SLUGS.append(_slug)
 
 # Every dataset all_hf pushes: default-species intervals versions + cohort slugs.
-ALL_DATASETS = list(INTERVALS_VERSIONS) + SPECIES_SUBSET_SLUGS
+ALL_DATASETS = [
+    iv for iv in INTERVALS_VERSIONS if iv not in CURATED_CCRE_SUBSETS
```
Comment (gonzalobenegas) [inline snakemake/zoonomia_projection_dataset/workflow/rules/dataset.smk:83]: 🤖 **Review finding — fixed in this commit.** Adding the curated subsets to `intervals_versions` would otherwise pull them into the bare `all_hf` target, diverging from `sky/upload.yaml`'s region-mode exclusion (a documented `snakemake -n all_hf` sanity check would then plan to build/upload the experimental probes). Now excluded here via the single-source `curated_ccre_subsets` config key; they still build via their explicit target / `UPLOAD_MODE=curated_ccre`.

File: src/marin_dna/pipelines/zoonomia_projection_dataset/region_labels.py [hunk 3]
```diff
@@ -487,6 +488,151 @@ def label_windows_bp_majority(
...
+    # naming convention. Cast both to ``str`` (the labels parquet uses bare
+    # Ensembl names like ``"1"``); a silent dtype/convention mismatch would zero
+    # all coverage and trip the assert below with a misleading message.
+    windows = (
+        base.select(["chrom", "start", "end"])
+        .to_pandas()
+        .astype({"chrom": str})
+        .reset_index(drop=True)
+    )
+    enhancer_df = enhancer_set.to_pandas().astype({"chrom": str})
```
Comment (gonzalobenegas) [inline src/marin_dna/pipelines/zoonomia_projection_dataset/region_labels.py:620]: 🤖 **Review finding — fixed here.** `_coverage_bp` matches windows↔region by chrom-key equality; the windows were cast to `str` but the cCRE region frames were not. A dtype/naming mismatch (categorical, or `chr1` vs `1`) would silently zero all coverage → all-`False` mask → the `0 windows` assert fires with a message blaming `cre_class`. Both region frames are now `.astype({"chrom": str})` and the assert names the chrom-convention cause.

File: snakemake/zoonomia_projection_dataset/sky/upload.yaml [hunk 4]
```diff
@@ -76,13 +82,21 @@ run: |
     cfg = yaml.safe_load(open('config/config.yaml'))
     pv = cfg['pipeline_version']
     mode = os.environ['UPLOAD_MODE']
+    # The #326 curated cCRE sub-filters are NOT region partitions; keep them out
+    # of region mode and give them their own curated_ccre mode. Single source of
+    # truth = config's curated_ccre_subsets (shared with dataset.smk + regions.smk),
+    # not a name-prefix heuristic that could catch an unintended future name.
+    curated = list(cfg.get('curated_ccre_subsets', []))
```
Comment (gonzalobenegas) [inline snakemake/zoonomia_projection_dataset/sky/upload.yaml:89]: 🤖 **Review finding — fixed here.** Was `[s for s in intervals_versions if s.startswith('v4_ccre_noexon')]` — a prefix heuristic that could silently catch an unintended future name (e.g. a later `v4_ccre_noexon_v2`) and misclassify it across region/curated modes, with no error since `test -n` only checks non-empty. Now reads the explicit `curated_ccre_subsets` config list (single source of truth shared with `dataset.smk` + `regions.smk`).

File: snakemake/zoonomia_projection_dataset/workflow/rules/dataset.smk [hunk 5]
```diff
@@ -260,6 +271,48 @@ rule dataset_hf_readme_v4:
         )
 
 
+# The curated-ccre card (below) and the v4 partition card both match v4_\w+;
+# prefer the curated card for the two #326 sub-filter names (the v4 card would
+# fail looking them up in the six-label composition TSV).
+ruleorder: dataset_hf_readme_curated_ccre > dataset_hf_readme_v4
```
Comment (gonzalobenegas) [inline snakemake/zoonomia_projection_dataset/workflow/rules/dataset.smk:277]: 🤖 **Review note (considered, kept).** Card routing for the curated subsets relies on this `ruleorder` + the exact-name `intervals_version` constraint on `dataset_hf_readme_curated_ccre` winning over `dataset_hf_readme_v4`'s broad `v4_\w+`. It is load-bearing (the v4 writer would `ValueError` on these names) but correct and dry-run-verified. A future `v4_ccre_*` dataset must be added to `curated_ccre_subsets` **and** this rule's constraint **and** `_CURATED_CCRE_BLURBS` — acceptable coupling for two fixed datasets, surfacing it so the next editor knows.

**Comment by gonzalobenegas** [comment]
🤖 **Datasets built + uploaded** (SkyPilot `curated_ccre` mode, r6i.8xlarge, ~9 min run on `6b320c2`). The pipeline planned exactly the two curated chains — **138 steps, no projection cascade** (the `--allowed-rules` collapse worked) — and `dataset_hf_readme_curated_ccre` produced the cards (not the v4 partition card).

| dataset | shards | rows (pre-RC) | samples (post-RC) | ≈ anchors |
|---|--:|--:|--:|--:|
| [`bolinas-dna/zoonomia-v1-v4_ccre_noexon`](https://huggingface.co/datasets/bolinas-dna/zoonomia-v1-v4_ccre_noexon) | 64 | 38,355,415 | 76,710,830 | ~372K |
| [`bolinas-dna/zoonomia-v1-v4_ccre_noexon_enhancer`](https://huggingface.co/datasets/bolinas-dna/zoonomia-v1-v4_ccre_noexon_enhancer) | 64 | 31,367,144 | 62,734,288 | ~304K |

**Invariants verified:** enhancer ⊂ noexon (31.4M < 38.4M, a strict subset by design); both < the `v4_ccre_non_promoter` baseline (44M pre-RC). The ~372K / ~304K anchor counts match the #283 H1-diagnostic estimates (~370K = 86.6%, ~304K = 71%) almost exactly — the filters behave as designed end-to-end. Cards render with the commit-pinned permalink + `biology/genomics/DNA` tags.

Ready to train (`experiments/exp326_curated_enhancers.py`, one `SWEEP_DATASETS` per arm).

**Comment by gonzalobenegas** [comment]
🤖 Closing without merging, per #340's merge policy (`main` = reusable-core; experiments/analysis stay on their branch). The curated cCRE datasets are built and live on HF (`bolinas-dna/zoonomia-v1-v4_ccre_noexon{,_enhancer}`), and the build code + exp326 training script remain permalinked from the #326 results. Branch `claude/issue-326-curated-enhancers` is the permanent reference.
